### ISSUE TITLE
User verbosity levels

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#325]
I tried an idea to make using link-parser (or the lib) more user friendly in case of problems, by enabling the user seeing more info regarding library "hidden" operations.
Will it be fine to **reserve verbosity values 1-3** for displaying user info?
(Programmer debug info will appear only on verbosity>=4, skipping the user info display.)

For example (please ignore the inconsistency of the missing program name from the dict messages - this can be fixed later):
```
$ link-parser -verbosity=2
verbosity set to 2
Info: Current directory: /usr/local/src/link-grammar-devel/windows/data/en
Info: Last-resort data directory: NULL
Info: Opening file ./en/dict.db (Not found)
Info: Opening file ./data/en/dict.db (Not found)
Info: Opening file ../en/dict.db (Not found)
Info: Opening file ../data/en/dict.db (Not found)
Info: Opening file /usr/local/share/link-grammar/en/dict.db (Not found)
Info: Opening file en/dict.db (Not found)
Info: Opening file ./en/4.0.dict (Not found)
Info: Opening file ./data/en/4.0.dict (Not found)
Info: Opening file ../en/4.0.dict
Info: Opening file ../en/4.0.regex
Info: Opening file ../en/4.0.knowledge
Info: Opening file ../en/4.0.constituent-knowledge
link-grammar: Info: Dictionary found at ../en/4.0.dict
link-grammar: Info: Using locale LC_CTYPE=en_US.utf8 ...
link-grammar: Info: Dictionary version 5.3.7.
link-grammar: Info: Library version link-grammar-5.3.7. Enter "!help" for help.
>
```

**Comment by linas** [comment]
Sounds fine to me. But ...

The stark reality is that there are very few users of LG, and I find it very unlikely that any of them will ever play with the debug flags or attempt to do any significant debugging at all, with them.  You would be the only user of this feature; I might try it on rare occasions, but I am used to just adding printf to the code I am interested in.

If you are casting about for something to do, I would still encourage you to delve deeper into the theory.  The most important thing lacking from LG today is an ability to have it learn new grammars for new languages. Probably little else matters.

**Comment by ampli** [comment]
It may be there are very few users of LG because it is not clear what can be done with it, and there is a steep learning curve involving using it, partially because not up-to-date and missing documentation, and also missing examples. My suggestion to add user verbosity levels is not for real debug but for helping users to see what went wrong when the program doesn't work for them (e.g. not finding their dictionary in the case I demonstrated). Of course these also needs to be documented.

Another problem is that most of the users use Windows, so providing pre-compiled Windows binaries for each release may help. I would also like to make it much easier to compiler the library code on Windows. For example - the regex library requirement can be easily removed by providing a C++ file which emulates the POSIX regex API by C++11 regex library calls. An option for line editing can be added too.

I also think that making a major revision of the documentation may help in this regard.
I would also like to add example program that use the library.
For example, a program to fix "bad grammar". This has added benefits [that I omit here]...

Regarding my ongoing work on the library: Because I invested time in understanding the code, I know how to fix many problems in it. For example to speed it up: a pruning hack made Russian parsing twice faster - providing the exact same results of course (but there is a missing infrastructure to do it cleanly). A SAT change in generating XOR expressions made parsing long sentences 50% faster (in the queue for final integration), and I still investigate if similar changes in generating AND/OR expressions is possible. It is possible to fix the "cost" in SAT. And more.
I think that it could be a major achievement if the parsing could be done several times faster (may help automatically generated dictionaries that can be very complex).
I also thought on implementing other things that appear in the README todo list.

Regarding theory, of course I think very much on that, and have several ideas in that regard. However, I think it will be better if I first provide proof-of-concept programs for each of the ideas.
I also thought of course on tokenizing non-concatenative languages, and found this should be done in a different way concatenetive tokenization is done. But without an actual proof-of-concept these are only words...
